docs-src/architecture/state-management.md

```diff
@@ -0,0 +1,93 @@
+
+Q2: How memory grows based on this behaviour?
+
+When 4α runs, and the document cache contains 3 entries (2,3,4), in total we have 4 document states, since task 1α has not released document 1 state to garbage collector. Let us call this a _spawned_ task - the one that operates on a document that is no longer in cache. When a new task appears for a new document, we will end up with one more spawned task with one more instance of the document state residing in memory. This way memory grows based on how much concurrent operations we have. To bound this growth, we added a configuration parameter `concurrentTasksLimit`. This is a maximum number of concurrently running tasks for all the documents we have. When this limit is reached, the next task has to wait till one of the tasks is done.
+
+One could consider ``concurrentTasksLimit` as a ceiling for another vector of memory growh, compared to the cache limit. This vector is related to bursts of document processing.
+
+### Subscriptions and two parts of cache
+
+One of the features is ability to subscribe to document changes. When a consumer subscribes to doctype updates, the updates should flow regardless of the pinning status. To enable that in the presence of the cache eviction, we mandate that a subscribed document is not evicted. For this, we internally split the cache into two buckets: *volatile*, and *durable*. Volatile is a vanilla LRU map, with cache eviction. Durable bucket maintains list of subscribed documents. When document is subscribed, we put the state in durable bucket. When unsubscribed, it moves to volatile bucket. Interesting question here is what state to put to the durable bucket. We have few scenarios here:
+
+- document is in memory ⇒ we use in-memory state
+- document is pinned but not in memory ⇒ we load from the state store,
+- document is not pinned and not in memory ⇒ we use state from Doctype.
+
+So, at any time, we are sure we start subscription with the latest available state.
+
+### Doctype and State
+
+As part of state refactor effort we have decided to maintain a certain semantics for how Doctype updates state. After created, Doctype maintains just the state it was initialized with. It only updates state after being changed or when subscribed. In the latter case, the state gets continuously updated. If looked from the inside, one could notice a Doctype relies on `RunningStateLike` instance to maintain the state. One could see `RunningStateLike` as either an Observable of DocState with an access to the current value, or as BehaviorSubject with some additional getters. The reason for this, is we want to avoid explicit memory management in Doctype instances. Any way `RunningStateLike` is a kind of Subject, that has to be explicitly closed, and as one could see from sections above managing document state life cycle is complicated. Ceramic instance maintains that closing behaviour, and that frees a developer from closing every Doctype manually.
```

Comment:
```suggestion
As part of state refactor effort we have decided to maintain a certain semantics for how Doctype updates state. After created, Doctype maintains just the state it was initialized with. It only updates state after being manually changed or when subscribed. In the latter case, the state gets continuously updated. If looked from the inside, one could notice a Doctype relies on `RunningStateLike` instance to maintain the state. One could see `RunningStateLike` as either an Observable of DocState with an access to the current value, or as BehaviorSubject with some additional getters. The reason for this, is we want to avoid explicit memory management in Doctype instances. Any way `RunningStateLike` is a kind of Subject, that has to be explicitly closed, and as one could see from sections above managing document state life cycle is complicated. Ceramic instance performs that closing behavior when the CeramicAPI instance itself is closed, and that frees a developer from closing every Doctype manually. However, the more open and subscribed Doctype instances there, the more memory is used, as having lots of subscribed Doctype instances can cause the in-memory cache to grow behind the configured max cache size.
```